application/states.go

```diff
@@ -0,0 +1,228 @@
+	currentDataset *models.DatasetUpdate, // Called Dataset in Mongo
+	currentVersion *models.Version, // Called Instances in Mongo
+	versionUpdate *models.Version, // Next version, that is the new version
+	versionDetails VersionDetails) error {
+	fmt.Println("Edition confirming")
+	return nil
+}
+func (g EditionConfirmed) String() string {
+	return "edition-confirmed"
+}
+
+type Detached struct{}
+
+func (g Detached) Enter(combinedVersionUpdate *models.Version, l *StateMachine, smDS *StateMachineDatasetAPI, ctx context.Context,
+	currentDataset *models.DatasetUpdate, // Called Dataset in Mongo
+	currentVersion *models.Version, // Called Instances in Mongo
+	versionUpdate *models.Version, // Next version, that is the new version
+	versionDetails VersionDetails) error {
+	fmt.Println("Detaching")
+	return nil
+}
+func (g Detached) String() string {
+	return "detached"
+}
+
+type Associated struct{}
+
+func (g Associated) Enter(combinedVersionUpdate *models.Version, l *StateMachine, smDS *StateMachineDatasetAPI, ctx context.Context,
+	currentDataset *models.DatasetUpdate, // Called Dataset in Mongo
+	currentVersion *models.Version, // Called Instances in Mongo
+	versionUpdate *models.Version, // Next version, that is the new version
+	versionDetails VersionDetails) error {
+	fmt.Println("Associating")
+	err := models.ValidateVersion(combinedVersionUpdate)
+	if err != nil {
+		fmt.Println("Validation Failed")
+		fmt.Println(err)
+		return err
+	} else {
+		fmt.Println("Validation passed, continue")
+		return nil
+	}
+}
+func (g Associated) String() string {
+	return "associated"
+}
+
+type Published struct{}
+
+func (g Published) Enter(combinedVersionUpdate *models.Version, l *StateMachine, smDS *StateMachineDatasetAPI, ctx context.Context,
+	currentDataset *models.DatasetUpdate, // Called Dataset in Mongo
+	currentVersion *models.Version, // Called Instances in Mongo
+	versionUpdate *models.Version, // Next version, that is the new version
+	versionDetails VersionDetails) error {
+	fmt.Println("Publishing")
+
+	// This needs to do the validation on required fields etc.
+	errModel := models.ValidateVersion(combinedVersionUpdate)
+	if errModel != nil {
+		fmt.Println("Validation Failed")
+		fmt.Println(errModel)
+		return errModel
+	} else {
+		fmt.Println("Validation passed, continue")
+		eTag := headers.IfMatchAnyETag
+		if currentVersion.ETag != "" {
+			eTag = currentVersion.ETag
+		}
+
+		lockID, error := smDS.dataStore.Backend.AcquireInstanceLock(ctx, currentVersion.ID)
+		if error != nil {
+			return error
+		}
+		defer func() {
+			smDS.dataStore.Backend.UnlockInstance(ctx, lockID)
+		}()
+
+		if _, errVersion := smDS.dataStore.Backend.UpdateVersion(ctx, currentVersion, combinedVersionUpdate, eTag); errVersion != nil {
+			return errVersion
+		}
+
+		data := versionDetails.baseLogData()
+		log.Info(ctx, "attempting to publish version", data)
```

Comment:
I believe the `UpdateVersion` logic above will have already changed the state on the version to published in the DB. Therefore, the log line should probably be above the update call.

```suggestion
		data := versionDetails.baseLogData()
		log.Info(ctx, "attempting to publish version", data)
		
		if _, errVersion := smDS.dataStore.Backend.UpdateVersion(ctx, currentVersion, combinedVersionUpdate, eTag); errVersion != nil {
			return errVersion
		}
```

Reply:
The log line has now been amended